## Stage 1: Installing dependencies and setting up GPU environment

In [1]:
!pip uninstall tensorflow


Uninstalling tensorflow-2.2.0:
  Would remove:
    /usr/local/bin/estimator_ckpt_converter
    /usr/local/bin/saved_model_cli
    /usr/local/bin/tensorboard
    /usr/local/bin/tf_upgrade_v2
    /usr/local/bin/tflite_convert
    /usr/local/bin/toco
    /usr/local/bin/toco_from_protos
    /usr/local/lib/python3.6/dist-packages/tensorflow-2.2.0.dist-info/*
    /usr/local/lib/python3.6/dist-packages/tensorflow/*
Proceed (y/n)? y
  Successfully uninstalled tensorflow-2.2.0


In [2]:
pip install tensorflow-gpu==2.0.0-alpha0


     |████████████████████████████████| 332.1MB 27kB/s 
     |████████████████████████████████| 419kB 30.1MB/s 
     |████████████████████████████████| 3.0MB 37.6MB/s 


## Stage 2: Import dependencies for the project

In [3]:
import numpy as np
import datetime
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist

/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:523: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:524: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or 

In [4]:
tf.__version__

'2.0.0-alpha0'

## Stage 3: Dataset preprocessing



### Loading the dataset

In [0]:
#Loading the Fashion Mnist dataset
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

### Image normalization

We devide each image in the training and testing dataset with the maxiumum number of pixels (255).

In this way each pixel will be in the rainge [0, 1]. By normalizing imaes we are making sure that our model (ANN) trains faster.

In [0]:
X_train = X_train / 255.0

In [0]:
X_test = X_test / 255.0

### Reshaping of the dataset



In [0]:
#Since each image is 28x28, we simply use reshape the full dataset to [-1 (all elements), height * width]
X_train = X_train.reshape(-1, 28*28)

In [9]:
X_train.shape

(60000, 784)

In [0]:
#Reshape the testing subset in the same way
X_test = X_test.reshape(-1, 28*28)

## Stage 4: Building an Artificial Neural network

### Defining the model



In [0]:
model = tf.keras.models.Sequential()

### Adding the first layer (Dense layer)

Layer hyper-parameters:
- number of units/neurons: 128
- activation function: ReLU
- input_shape: (784, )

In [0]:
model.add(tf.keras.layers.Dense(units=128, activation='relu'))

### Adding a Dropout layer 



In [0]:
model.add(tf.keras.layers.Dropout(0.4))

### Adding the second layer (output layer)

- units == number of classes (10 in the case of Fashion MNIST)
- activation = 'softmax'

In [0]:
model.add(tf.keras.layers.Dense(units=10, activation='softmax'))

### Comiling the model

- Optimizer: Adam
- Loss: Sparse softmax (categorical) crossentropy 

In [0]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['sparse_categorical_accuracy'])

In [37]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 256)               200960    
_________________________________________________________________
dropout (Dropout)            (None, 256)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                2570      
_________________________________________________________________
dense_2 (Dense)              (None, 128)               1408      
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1

### Training the model

In [38]:
model.fit(X_train, y_train, epochs=10)

Epoch 1/10
60000/60000 [==============================] - 12s 208us/sample - loss: 0.4980 - sparse_categorical_accuracy: 0.8771
Epoch 2/10
60000/60000 [==============================] - 12s 203us/sample - loss: 0.3273 - sparse_categorical_accuracy: 0.8922
Epoch 3/10
60000/60000 [==============================] - 12s 204us/sample - loss: 0.3133 - sparse_categorical_accuracy: 0.8947
Epoch 4/10
60000/60000 [==============================] - 12s 207us/sample - loss: 0.3085 - sparse_categorical_accuracy: 0.8956
Epoch 5/10
60000/60000 [==============================] - 12s 203us/sample - loss: 0.2964 - sparse_categorical_accuracy: 0.8983
Epoch 6/10
60000/60000 [==============================] - 12s 207us/sample - loss: 0.2969 - sparse_categorical_accuracy: 0.8975
Epoch 7/10
60000/60000 [==============================] - 12s 202us/sample - loss: 0.2876 - sparse_categorical_accuracy: 0.9009
Epoch 8/10
60000/60000 [==============================] - 12s 208us/sample - loss: 0.2802 - sparse_categ

### Model evaluation and prediction

In [39]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)

10000/10000 [==============================] - 1s 95us/sample - loss: 0.3548 - sparse_categorical_accuracy: 0.8815
